File: email.ts
/**
 * this module provides normalization and validation of email address.
 * @module
 */
export const EMAILPattern: string = '^(?:"?([^"]*)"?\\s)?(?:<?(.+@[^>]+)>?)$';
export const reEMAIL: RegExp = new RegExp(EMAILPattern);

/**
 * normalize themail address.
 * @param email 
 * @returns 
 * 
 * @example
 * ```ts
 *  const email: string = normoralizeEMAIL('  jane.doe@example.edu ');
 *  console.log(email); // 'jane.doe@example.edu'
 * ```
 */
export function normalizeEMAIL(email: string): string {
  return email.trim().replaceAll(/\s+/g, "").trim();
}

/**
 * validates the form of the email address after normalizing it.
 * @param email 
 * @returns 
 * 
 * @example
 * ```ts
 *  const email: string = normoralizeEMAIL('  jane.doe@example.edu ');
 *  if (validateEMAIL(email) {
 *      console.log(`${email} appears to be valid`);
 *  } else {
 *      console.error(`${email} fails validation`);
 *  }
 * ```
 */
export function validateEMAIL(email: string): boolean {
  const normalized = normalizeEMAIL(email);
  return reEMAIL.test(normalized);
}

File: email.go
package metadatatools

import (
	"regexp"
	"strings"
)

// EMAILPattern is the regular expression pattern matching a normalized
// email address.
const EMAILPattern = `^(?:"?([^"]*)"?\s)?(?:<?(.+@[^>]+)>?)$`

// reEMAIL is the compiled form of EMAILPattern.
var reEMAIL = regexp.MustCompile(EMAILPattern)

// NormalizeEMAIL trims whitespace from email and removes any internal
// whitespace.
//
// Example:
//
//	email := NormalizeEMAIL("  jane.doe@example.edu ")
//	// email == "jane.doe@example.edu"
func NormalizeEMAIL(email string) string {
	return strings.Join(strings.Fields(email), "")
}

// ValidateEMAIL validates the form of email after normalizing it.
func ValidateEMAIL(email string) bool {
	return reEMAIL.MatchString(NormalizeEMAIL(email))
}
